File: src/content/docs/topics/cs/coding-problems/greedy/134-gas-station.py
def can_complete_circuit(gas: list[int], cost: list[int]) -> int:
    pass  # TODO: implement

def _run_tests() -> None:
    assert can_complete_circuit([1, 2, 3, 4, 5], [3, 4, 5, 1, 2]) == 3
    assert can_complete_circuit([2, 3, 4], [3, 4, 3]) == -1
    assert can_complete_circuit([1], [1]) == 0
    assert can_complete_circuit([5], [4]) == 0
    assert can_complete_circuit([0], [0]) == 0
    assert can_complete_circuit([10_000], [10_000]) == 0
    assert can_complete_circuit([2, 0, 1], [0, 1, 2]) == 0

    n = 100_000
    large_gas = [0] * n
    large_cost = [0] * n
    for i in range(10_000):
        large_cost[i] = 1
    large_gas[10_000] = 10_000

    import time as _t
    _t0 = _t.perf_counter()
    assert can_complete_circuit(large_gas, large_cost) == 10_000
    _ms = (_t.perf_counter() - _t0) * 1000
    print(f'perf can_complete_circuit(n=100000): {_ms:.1f}ms')
    print('all tests pass')

if __name__ == '__main__':
    _run_tests()

File: src/content/docs/topics/cs/coding-problems/greedy/134-gas-station.swift
// LEETCODE_TYPE: Solution
func expectEqual<T: Equatable>(
    _ actual: T,
    _ expected: T,
    _ message: String = "",
    file: StaticString = #fileID,
    line: UInt = #line
) {
    guard actual == expected else {
        let detail = message.isEmpty ? "values differ" : message
        fatalError("\(file):\(line): \(detail). Expected \(expected), got \(actual)")
    }
}

func expectTrue(
    _ condition: @autoclosure () -> Bool,
    _ message: String = "expected true",
    file: StaticString = #fileID,
    line: UInt = #line
) {
    guard condition() else {
        fatalError("\(file):\(line): \(message)")
    }
}

func reportSuccess() {
    print("All Swift tests passed")
}
final class Solution {
    func canCompleteCircuit(_ gas: [Int], _ cost: [Int]) -> Int {
        // TODO: Implement
        fatalError("TODO: Implement")
    }
}

func runTests() {
    // TEST_VECTORS_BEGIN sha256:cf38ec67d4b22e2f3736614bea688390e3d7ca7d414a985e3bf05a180ee348e4
    expectEqual(Solution().canCompleteCircuit([1, 2, 3, 4, 5], [3, 4, 5, 1, 2]), 3, "starts-at-three")
    expectEqual(Solution().canCompleteCircuit([2, 3, 4], [3, 4, 3]), -1, "no-valid-start")
    expectEqual(Solution().canCompleteCircuit([5], [4]), 0, "single-station")
    // EXCLUDED_VECTOR mismatched-lengths: [[1,2],[1]] | Gas and cost arrays must have the same nonzero length.
    // TEST_VECTORS_END
    reportSuccess()
}

runTests()
